docs/design/build-system.md

```diff
@@ -0,0 +1,154 @@
+These files are not processed and commonly include images, favicons, pdfs and other generic file types.
+They are loaded into the compilation using `copy-webpack-plugin`.
+
+
+### Scripts
+
+Scripts in the build system refer specifically to Javascript files that are meant to be loaded directly on `index.html` without being processed.
+They are loaded into the compilation using a custom webpack plugin.
+
+
+### Third party dependencies
+
+Third party dependencies are mostly inside `node_modules` and are referenced via imports in source files.
+Stylesheet third party dependencies are treated mostly the same as sources.
+
+JavaScript third party dependencies suffer a more involved process.
+They are first resolved to a folder in `node_modules` via [Node Module Resolution](https://nodejs.org/api/modules.html#modules_modules).
+Afterwards one of the entry points in that folder's `package.json` is used following the `es2015 > browser > module > main` priority list.
+
+Once the actual JavaScript file is determined, it is loaded into the compilation together with it's source map.
+
+
+## Code splitting
+
+Code is automatically split into different files (or chunks, for js files) based on a few different triggers.
+
+The main TypeScript entry point and it's dependencies are bundled into the `main` chunk.
+Global styles and scripts get one file per entry, named after themselves.
+
+JavaScript code imported only via dynamic imports is automatically split into a separate chunk that is loaded asynchronously named after the file containing the dynamic import.
+If multiple asynchronous chunks contain a reference to the same module, it is placed in a new asynchronously loaded chunk named after the other chunks that use it.
+
+There is also a special chunk called `runtime` that contains the module loading logic and is loaded before the others.
+
+
+## Optimizations
+
+The build system contains optimizations aimed and improving the performance, for development builds, or the size of artifacts, for production builds.
+These are often mutually exclusive and thus we cannot just default to always using them.
+
+
+### Development optimizations
+
+Development optimizations focus on reducing rebuild time on watched builds.
+Although faster is always better, our threshold is to keep rebuilds even for large projects below 2 seconds.
+
+Computation needed to bundle code grows with its total size because of the cost of string concatenation and source map operations.
+Third party dependencies that are initially loaded are split into a synchronously loaded chunk called `vendor`.
+Splitting the infrequently changed vendor code from the frequently changed source code thus helps make rebuilds faster.
+
+When processing stylesheets Webpack stores the intermediate modules as javascript code.
+The javascript wrapper code makes stylesheets larger and `mini-css-extract-plugin` must be used to obtain the actually stylesheet content into a CSS file.
```

Comment:
Nit: Should be consistent on the capitalization of "JavaScript" at least throughout this doc.

Reply:
Fixed